msim/src/sim/net/network.rs

```diff
@@ -195,6 +199,50 @@ impl Network {
         Ok(addr)
     }
 
+    pub fn register_tcp_id(&mut self, node: NodeId, tcp_id: u32) {
+        trace!("registering tcp id {} for node {}", tcp_id, node);
+        assert!(
+            self.nodes
+                .get_mut(&node)
+                .unwrap()
+                .live_tcp_ids
```

Comment:
How does this differ from `Endpoint::live_tcp_ids`?

Reply:
it is semi-duplicated, but it has to be - or at least I don't see an obvious simplification.

Endpoint::live_tcp_ids tracks the ownership/registration of the ids. This is required so that when the endpoint is dropped we can verify that we cleaned everything up.

Node::live_tcp_ids tracks all ids across the whole node, and may include ids owned by multiple endpoints. This is required so that we can tell whether the id of an incoming message still pertains to a live tcp connection. We can't use Endpoint::live_tcp_ids for this, because we want to distinguish between ConnectionRefused (no endpoint exists for socket address) and ConnectionReset (no such tcp id for node) errors.

In the case where the other side hung up, the endpoint won't exist any more, which currently indicates ConnectionRefused.